docs/DevGuide/TensorExpressions.md

```diff
@@ -0,0 +1,411 @@
+defined, but more spatial and spacetime indices (letters) can easily be added if
+needed
+
+The properties of each `TensorIndex` and the `Tensor`'s indices (typelist of
+\ref SpacetimeIndex "TensorIndexType"s) must be compatible:
+- valences must match
+- if a `Tensor`'s index is spacetime, you can use a spacetime `TensorIndex`,
+spatial `TensorIndex`, or concrete time `TensorIndex`
+- if a `Tensor`'s index is spatial, you must use a spatial `TensorIndex`
+
+To demonstrate correct and incorrect usage, let's say we have tensors
+\f$R_{ab}\f$ (two spacetime indices, e.g. type \ref tnsr "tnsr::ab") and
+\f$S_{ij}\f$ (two spatial indices, e.g. type \ref tnsr "tnsr::ij"):
+
+```
+R(ti::c, ti::d) // OK
+R(ti::c, ti::k) // OK, can use spatial TensorIndex on spacetime index
+R(ti::c, ti::t) // OK, can use time TensorIndex on spacetime index
+R(ti::c, ti::D) // ERROR, ti::D is upper but the 2nd index is lower
+
+S(ti::j, ti::k) // OK
+S(ti::a, ti::k) // ERROR, can't use spacetime TensorIndex on a spatial index
+S(ti::j, ti::t) // ERROR, can't use time TensorIndex on a spatial index
+```
+
+# Examples {#te_examples}
+## Basic operations {#te_basic_operations}
+
+In the following examples:
+- `R` is type \ref tnsr "tnsr::ab<DataVector, 3>"
+- `S` is type \ref tnsr "tnsr::ab<DataVector, 3>"
+- `T` is type \ref Scalar "Scalar<DataVector>"
+- `U` is type \ref tnsr "tnsr::Ab<DataVector, 3>"
+- `V` is type \ref tnsr "tnsr::aBC<DataVector, 3>"
+- `G` is type \ref tnsr "tnsr::a<DataVector, 3>"
+- `H` is type \ref tnsr "tnsr::A<DataVector, 3>"
+
+### Addition and subtraction {#te_addition_and_subtraction}
+\f$L_{ab} = R_{ab} + S_{ba}\f$
+```
+auto L = tenex::evaluate<ti::a, ti::b>(R(ti::a, ti::b) + S(ti::b, ti::a));
+```
+\f$L = 1 - T\f$
+```
+auto L = tenex::evaluate(1.0 - T());
+```
+
+### Contraction of a single tensor {#te_contraction}
+\f$L = U^{a}{}_{a}\f$
+```
+auto L = tenex::evaluate(U(ti::A, ti::a));
+```
+\f$L^b = V_{a}{}^{ba}\f$
+```
+auto L = tenex::evaluate<ti::B>(V(ti::a, ti::B, ti::A));
+```
+
+### Inner and outer products {#te_products}
+\f$L = G_a H^{a}\f$
+```
+auto L = tenex::evaluate(G(ti::a) * H(ti::A));
+```
+\f$L_{cb} = T G_a G_c U^{a}{}_{b}\f$
+```
+auto L =
+    tenex::evaluate<ti::c, ti::b>(T() * G(ti::a) * G(ti::c) * U(ti::A, ti::b));
+```
+
+### Division {#te_division}
+\f$L_a = \frac{G_a}{2}\f$
+```
+auto L = tenex::evaluate<ti::a>(G(ti::a) / 2.0);
+```
+\f$L_{ba} = \frac{R_{ab}}{T}\f$
+```
+auto L = tenex::evaluate<ti::b, ti::a>(R(ti::a, ti::b) / T());
+```
+\f$L = \frac{5}{U^{a}{}_{a} + 1}\f$
+```
+auto L = tenex::evaluate(5.0 / (U(ti::A, ti::a) + 1.0));
+```
+
+### Square root {#te_square_root}
+\f$L = \sqrt{T}\f$
+```
+auto L = tenex::evaluate(sqrt(T()));
+```
+\f$L = \sqrt{G_a H^a}\f$
+```
+auto L = tenex::evaluate(sqrt(G(ti::a) * H(ti::A)));
+```
+
+## More features {#te_more_features}
+
+### Assigning to a number {#te_assigning_to_a_number}
+You can assign a number (e.g. `double`) to a `Tensor` of any rank to fill all
```

Comment:
`Tensor<double>` so it's clear. I know you say it below, but since a reader sees this first, I think it's best to be explicit here.

Reply:
Ah, so this example shows you that you can do this for `Tensor<DataType>` though (kinda) in that it shows you how to do it for a `Tensor<double>` and then for a `Tensor<DataVector>`. It was meant to be like "here's what the syntax looks like for assigning to a number.... and here is what you have to do if you have a blaze vector type." But it does seem this was not clear, so let me know what could make it more clear

Reply:
Because the example wasn't even meant to demonstrate `Tensor<double>` specifically - I could have used `Tensor<std::complex<double>>` and done a one-liner, so the `Tensor` without tparams was purposeful

Reply:
Ah yeah I think explicitly splitting up the two cases would be good in your explanation.